runtime/compiler/aarch64/codegen/J9TreeEvaluator.cpp

```diff
@@ -1746,6 +1746,39 @@ static
 void genSuperClassTest(TR::Node *node, TR::Register *instanceClassReg, bool instanceClassRegCanBeReclaimed, TR::Register *castClassReg, int32_t castClassDepth,
                                             TR::LabelSymbol *falseLabel, TR_ARM64ScratchRegisterManager *srm, TR::CodeGenerator *cg)
    {
+
+   // When castClass is an unknown at compile time check if castClass is an interface or an array.
+   // If so skip superClassTest.
+   bool dynamicCastClass = castClassDepth == -1;
+   if ( dynamicCastClass )
+      {
+      TR::Register *scratchRegister1 = srm->findOrCreateScratchRegister();
+      TR_ASSERT(node->getSecondChild()->getOpCodeValue() != TR::loadaddr,
+            "genSuperClassTest: castClassDepth == -1 is not supported for a loadaddr castClass");
+
+      generateTrg1MemInstruction(cg, TR::InstOpCode::ldrimmx, node, scratchRegister1, TR::MemoryReference::createWithDisplacement(cg, castClassReg, offsetof(J9Class, romClass)));
+      generateTrg1MemInstruction(cg, TR::InstOpCode::ldrimmw, node, scratchRegister1, TR::MemoryReference::createWithDisplacement(cg, scratchRegister1, offsetof(J9ROMClass, modifiers)));
+
+      // Array an check
+      static_assert(J9AccClassArray == 0x10000, "J9AccClassArray must be 0x10000");
+      static_assert(J9AccInterface == 0x200, "J9AccInterface must be 0x200");
+      // Two cases can reach this code, an instanceOf node or a call to Class.cast() converted to a checkCast node
+      if (node->->getOpCodeValue() == TR::instanceof)
```

Comment:
Do you need this condition?
Instructions `tbnz reg, 16, falseLabel` and `tst reg, #0x10000 ; b.ne falseLabel` mean the same, except for the status of condition flags.

Reply:
I noticed in the code that instanceOf uses `tbz/tbnz` and checkCast uses `tst; b.ne`, I believe as the target for `tbz/tbnz` must be within 32KB and that doesn't work for checkCast, and tried to keep it the same. 
I guess I should ask if there's a benefit for `tbz/tbnz` to do it this way, or just use `tst; b.ne` for both cases here?

Reply:
`tbnz` is preferred when the destination is within the ±32KB range, because it is faster and it consumes less memory than `tst; b.ne` for this case.
Is the destination `falseLabel` far away for the checkCast case?

Reply:
Here is an example.  `genInstanceOfOrCheckCastObjectArrayTest()` takes a parameter `useTBZ` for deciding which code to use.

https://github.com/eclipse-openj9/openj9/blob/ae6f910e02511bd082b55b821036c435e00517e1/runtime/compiler/aarch64/codegen/J9TreeEvaluator.cpp#L1891-L1899

Reply:
It looks like even in the instanceOf case, the `falseLabel` can be a OOLCodeSection and can be placed far away. So we should use `tst; b.ne` for all cases.

Reply:
Corrected the code in b0255bc69f698fa27460f100006637d5808a45c9